NIFI-6388 Add dynamic relationships to the ExecuteScript processor.

kaHaleMaKai opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Add dynamic relationships to the ExecuteScriptProcessor.

The previous relationships `SUCCESS` and `FAILURE` are kept for backwards compability. The variable of type `Map<String, Relationship>` `rel` is introduced. If the processor has a dynamic property with name like `rel.my_new_relationship`, a relationship of name `my_new_relationship` is added.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ x ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ x ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ x ] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [ x ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ x ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ x ] Have you written or updated unit tests to verify your changes?
- [ x ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ x ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ x ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ x ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ x ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

mattyb149 commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/test/groovy/org/apache/nifi/processors/script/ExecuteScriptGroovyTest.groovy:
```diff
@@ -20,19 +20,17 @@ import org.apache.nifi.script.ScriptingComponentUtils
-import org.junit.BeforeClass
-import org.junit.Ignore
-import org.junit.Test
+import org.junit.*
```
Please ensure you've run the Maven build with the `-Pcontrib-check` profile activated, it will catch checkstyle errors such as star imports like this. Thanks!

mattyb149 replied:
Actually I don't think we have the Checkstyle rules for Groovy files in place, but just for consistency, it would be nice to avoid star imports where possible.

kaHaleMaKai replied:
You're right about it. Only the java code is checked. That's why I missed those imports. Will fix.

kaHaleMaKai replied:
fixed in 4614d29010d390d6ed5e50d56ec9fd3a29d55805

kaHaleMaKai commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java:
```diff
@@ -56,22 +59,31 @@
         + "Impact of sustained usage not yet verified.")
 @DynamicProperty(
-        name = "A script engine property to update",
+        name = "A script engine property to update, or a dynamic relationship",
```
will do, thanks

kaHaleMaKai commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java:
```diff
@@ -85,13 +97,30 @@
-    // Constants maintained for backwards compatibility
     public static final Relationship REL_SUCCESS = ScriptingComponentUtils.REL_SUCCESS;
     public static final Relationship REL_FAILURE = ScriptingComponentUtils.REL_FAILURE;
+    public static final String DYNAMIC_RELATIONSHIP_PATTERN_AS_STRING = "^rel\\.[a-zA-Z_][a-zA-Z0-9]*$";
```
No need for further restrictions. I had accepted variables names for most languages in mind (python, groovy etc) – and one could than get the relationship directly via `rel.my_relationship`.

But I'll relax the naming to the standard relationship names and prefix them with `REL_`. They'll be available in the corresponding hash map, indexed by string keys.

kaHaleMaKai commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java:
```diff
@@ -133,13 +167,57 @@
-                .build();
+                .dynamic(true);
+        if (isRelationship) {
+            builder.description(String.format(
```
great idea

kaHaleMaKai replied:
fixed in 5963a8e5576c527f035e0db15e0c95c07e413789

kaHaleMaKai commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java:
```diff
@@ -133,13 +167,57 @@
+        final boolean isRelationship = propertyDescriptorName != null && propertyDescriptorName.startsWith(DYNAMIC_RELATIONSHIP_PREFIX);
+        if (isRelationship) {
+            if (!DYNAMIC_RELATIONSHIP_PATTERN.matcher(propertyDescriptorName).matches()) {
+                log.warn("dyn. property for relationship is invalid: '{}'. accepted patterns: '{}'", new Object[]{propertyDescriptorName, DYNAMIC_RELATIONSHIP_PATTERN_AS_STRING});
```
fixed in 0b13c41acfe3f354db8dd739d929188ddb625a6a

kaHaleMaKai commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java:
```diff
@@ -264,4 +343,16 @@ public void onTrigger(ProcessContext context, ProcessSessionFactory sessionFacto
+            return new ValidationResult.Builder()
+                    .subject(subject)
+                    .input(input)
+                    .explanation("invalid dyn. relationship specified")
```
fixed in 0b13c41acfe3f354db8dd739d929188ddb625a6a

kaHaleMaKai commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java:
```diff
@@ -133,13 +167,57 @@
+                        .addValidator(new RelationshipInvalidator())
+                        .dynamic(true)
+                        .required(false)
+                        .name(propertyDescriptorName)
```
I tried that, but it didn't work out. We need to keep the full property name (including the `REL_` prefix, previously `rel.`), so that the `ExecuteScript#onPropertyModified()` method can detect the new property indicating a relationship. If we prematurely remove the prefix, the respective method cannot tell if it's just a regular dynamic property, or if the user wants to add a new dynamic relationship.

kaHaleMaKai commented:
@mattyb149 Thank you so far for your extensive feedback. Do you request any more changes? I will gladly provide them.

mattyb149 commented:
I'm waffling on whether we should have an explicit property (defaulted to false to retain behavior) to "Allow Custom Relationships". If it were set to true, then you'd apply the rules you have now (prefixed with `rel.`, e.g.), but if it's false, they are just treated as variables no matter the name.

Is that overkill for some obscure edge case (namely variables with periods in it, which aren't very Groovy but are technically usable)? @alopresto do you have any thoughts on this? I could go either way, just wanted to talk through it first. Otherwise this LGTM :)

kaHaleMaKai commented:
@mattyb149 Due to your suggestion, I have replaced the prefix `rel.` with `REL_` to honor the existing relationship naming (`REL_SUCCESS`, `REL_FAILURE`).

Besides from potential collisions with existing properties, the introduced variable `relationships` in the script context might already be in used by some scripts for whatever reason. Collisions are even more likely than.

I follow your suggestion and would like to add a flag for activating dynamic relationships (defaulting to false) to _opt-in_ into dynamic relationships. We could then deprecate the old behaviour and competely remove it (including the flag, if you like) in let's say NiFi 2.0. From then, the `REL_.*` named properties would always yield dynamic relationships no matter what. 

Thanks for naming the issue, I hadn't thought of that.

kaHaleMaKai commented:
@mattyb149 I submitted all proposed changes.

mattyb149 commented:
Sorry I lost track of this, do you mind rebasing against the latest master? Then I'll have a final look at it, thanks!

mattyb149 commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java:
```diff
@@ -91,13 +112,36 @@
+    protected void init(ProcessorInitializationContext context) {
+        super.init(context);
+        log = getLogger();
+        log.warn("The default for USE_DYNAMIC_RELATIONSHIPS will change from \"false\" to \"true\" starting in NIFI 2.0");
```
Let's not include this warning, in case we don't make the change in NiFi 2.0 we won't be providing false information

kaHaleMaKai replied:
done

mattyb149 commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java:
```diff
@@ -148,6 +196,87 @@ protected PropertyDescriptor getSupportedDynamicPropertyDescriptor(final String
+        if (!isValidRelationshipName(relName)) {
+            log.warn("dynamic property for relationship is invalid: '{}'. It must not be named REL_SUCCESS or REL_FAILURE (case in-sensitive)",
+                    new Object[]{propertyDescriptorName}
+            );
```
We should check for this in a validator set on the dynamic property, or in a `customValidate()` method. Since you have a validator, we shouldn't need the check or warning here, the framework should take care of marking the processor invalid if the relationship fails the validator. However that means you'd have to update the Validator to perform the check, rather than returning invalid as it does now.

kaHaleMaKai replied:
done

mattyb149 commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/script/ScriptingComponentUtils.java:
```diff
@@ -64,5 +68,15 @@
+    /** A property descriptor for specifying whether to use dynamic relationships or not */
+    public static final PropertyDescriptor USE_DYNAMIC_RELATIONSHIPS = new PropertyDescriptor.Builder()
+            .name("Use Dynamic Relationships")
+            .description("Whether properties prefixed with \"REL_\" should create a dynamic relationship or not. The default will change to \"true\" in NIFI 2.0.")
```
Let's delete the reference to changing the default, in case we don't make the change in NiFi 2.0 we won't be providing false information in the doc. Changing the default should be transparent to the user, because any existing instances in a flow migrated to NiFi 2.0 will retain the setting whether it is the default or not. Only new instances will have a different default, and can be changed at that time by the user. It would probably be worth mentioning in any migration document when/if the default is changed, just for completeness, but the impact should be negligible.

kaHaleMaKai replied:
done

mattyb149 commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/script/ScriptingComponentUtils.java:
```diff
@@ -64,5 +68,15 @@
+    public static final PropertyDescriptor USE_DYNAMIC_RELATIONSHIPS = new PropertyDescriptor.Builder()
+            .name("Use Dynamic Relationships")
+            .description("Whether properties prefixed with \"REL_\" should create a dynamic relationship or not. The default will change to \"true\" in NIFI 2.0.")
+            .required(false)
```
The convention for properties that use AllowableValues is to make them `required(true)`, not a big deal here since it will always be populated, but I mention it for consistency with other processors/properties.

kaHaleMaKai replied:
done

mattyb149 commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/test/groovy/org/apache/nifi/processors/script/ExecuteScriptGroovyTest.groovy:
```diff
@@ -16,23 +16,20 @@
 import java.util.concurrent.TimeUnit
 
-import static org.junit.Assert.assertNotNull
+import static org.junit.Assert.*
```
We don't have CheckStyle rules being applied to Groovy code at the moment, but we discourage the use of star imports (and even prohibit them in Java code via a CheckStyle rule), so would be good to have individual imports here.

kaHaleMaKai replied:
done

kaHaleMaKai commented:
If required, I'll rebase the code against 1.12-SNAPSHOT, after 1.11 gets released.

kaHaleMaKai commented:
Hi @mattyb149,

I have implemented all requested changes and rebased against 1.12-snapshot.
Please consider my PR again.

Thanks in advance!
Best,
Lars

mattyb149 commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java:
```diff
@@ -61,23 +64,41 @@
 @DynamicProperty(
-        name = "A script engine property to update",
-        value = "The value to set it to",
+        name = "A script engine property to update, or a dynamic relationship (if the property starts with \"REL_\" and USE_DYNAMIC_RELATIONSHIPS is set to true). "
```
This is only my opinion and offered for consistency's sake (behind the scenes), it goes along with some of my other comments. Why not make the name of the property the bound variable name and the value the relationship name? Meaning I could add a property `REL_TEST` with value `test` and use `REL_TEST` in the script but route `test` to downstream processors? This isn't a hard requirement from me but I think it would fit in well for folks who have written full processors as well as scripts for ExecuteScript.

mattyb149 commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java:
```diff
@@ -61,23 +64,41 @@
-        value = "The value to set it to",
+        name = "A script engine property to update, or a dynamic relationship (if the property starts with \"REL_\" and USE_DYNAMIC_RELATIONSHIPS is set to true). "
+                + "For dynamic relationships, the prefix \"REL_\" will be removed from the property name",
+        value = "The value to set it to. For a dynamic relationship, the value will become the documentation of that relationship",
```
I certainly applaud your eye toward being able to provide documentation around dynamic properties. However in this case (see my comment above) it might be more natural to use the value as the relationship name (displayed to the user). If they want full documentation and control over the relationship properties, they can use InvokeScriptedProcessor.

mattyb149 commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java:
```diff
@@ -61,23 +64,41 @@
+@DynamicRelationship(
+        name = "A relationship to add",
+        description = "If a dynamic property starts with \"REL_\" and USE_DYNAMIC_RELATIONSHIPS is set to true, it is assumed to be the name of a dynamic "
+        + "relationship to add. In that case, the prefix \"REL_\" will be removed from the actual name of the "
```
If you change the use of the dynamic relationships' names and properties, the doc should be updated as well to reflect that behavior.

mattyb149 commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java:
```diff
@@ -216,14 +329,21 @@ public void onTrigger(ProcessContext context, ProcessSessionFactory sessionFacto
                 bindings.put("REL_FAILURE", REL_FAILURE);
+                // only add map relationships if user has opted in
+                if (this.useDynamicRelationships) {
+                    bindings.put("relationships", getRelationshipsAsMap());
```
What about adding the dynamic relationships the same way as REL_SUCCESS and REL_FAILURE? We wouldn't need an extra map if we could just `session.transfer(flowFile, REL_TEST)`, and it would be more consistent with the usage of the existing two relationships. You could put the code below into an `else` clause for `if(this.useDynamicRelationships)`, that way you add a REL_XYZ binding in either case, but it's either set to the Relationship `XYZ` or the PropertyDescriptor `XYZ` depending on the setting of "Use Dynamic Relationships"?

mattyb149 commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java:
```diff
@@ -303,4 +423,57 @@ public void stop() {
+                || REL_FAILURE.getName().equals(nameInLowerCase));
+    }
+
+    private static class RelationshipInvalidator implements Validator {
```
This is a nitpick, but I'm not sure we have any other "Invalidator" implementations, usually they're called "Validator" even if their only job is to check conditions that may make it invalid.

mattyb149 commented on nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/test/resources/groovy/testDynamicRelationships.groovy:
```diff
@@ -0,0 +1,23 @@
+ */
+def flowFile = session.create()
+flowFile = session.putAttribute(flowFile, "time-updated", new Date().toString())
+session.transfer(flowFile, relationships["test"])
```
If the above suggestions are implemented, this is just a note to remind that the `relationships` map would be removed and you could use REL_TEST inline instead.
